sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala

```diff
@@ -2755,6 +2755,20 @@ class DataFrameSuite extends QueryTest
     )
     checkAnswer(test.select($"best_name.name"), Row("bob") :: Row("bob") :: Row("sam") :: Nil)
   }
+
+  test("SPARK-34829: Typed ScalaUDF result conversion works") {
+    val reverse = udf((s: String) => s.reverse)
+    val df = Seq(Array("abc", "def")).toDF("array")
+    val test = df.select(transform(col("array"), s => reverse(s)))
+    checkAnswer(test, Row(Array("cba", "fed")) :: Nil)
```

Comment:
Can this issue happens only in codegen paths? Could you add tests for interpreted paths, too?

Reply:
I don't think there is a codegen path for `ArrayTransform`. It can call `ScalaUDF.eval()` on its interpreted path (https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala#L282) only.

Reply:
Updated description as well.